## Introduction:

Reference: 
https://stackabuse.com/implementing-svm-and-kernel-svm-with-pythons-scikit-learn/

You can practise more about SVM through the following link:

https://www.geeksforgeeks.org/classifying-data-using-support-vector-machinessvms-in-python/

### set of imports .. 

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split  #data_split
from sklearn.svm import SVC  #SVM
from sklearn.metrics import classification_report, confusion_matrix  #model eval

%matplotlib inline

In [2]:
bankdata = pd.read_csv("bill_authentication.csv")
bankdata.head()

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

Here we have 4 indepedent variables and single dependent variable (Class) that we gonna classifiy using SVM ..

In [8]:
X = bankdata.iloc[:, :-1].values
y = bankdata.iloc[:, 4].values

Let's check possible value of the class variable .. 

In [9]:
bankdata['Class'].unique()

array([0, 1], dtype=int64)

all possible values of the class .. either 0 or 1

### Data Split

Here we gonna use 80-20 split .. 

however, you might use another type (i.e. 70-30, or 10-fold cross validation) .. 

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20)

## Modeling using SVM:

- We gonna utilize the built in function entitled SVC() from sklearn library
- This is a model training phase using the funtion fit()

In [12]:
svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='linear', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

Here we use "linear" kernal .. meanwhile, you might use non-linear kernal if you have more complex dataset to obtain better results .. 

You can read more about SVM through the following link .. 

https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

### Prediction phase:

This is the model testing phase .. using the function predict() 

In [14]:
y_pred = svclassifier.predict(X_test)
y_pred

array([0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1], dtype=int64)

These are the class values corresponding to each input in the test dataset .. 

Did our model perform well ?! .. 

Let's evalaute it ! 

Usually, in regression tasks we evaluate the models via calculating the error (i.e. mean squared error) .. 
and in classification tasks .. the performance of the algorithm is done using confusion martrix and accuracy ..

Hence, let's check the confusion matrix of our SVM algorithm .. 

In [18]:
print(confusion_matrix(y_test,y_pred))

[[149   2]
 [  0 124]]


In [19]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.99      0.99       151
           1       0.98      1.00      0.99       124

   micro avg       0.99      0.99      0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275



Perfect! .. 

The model performed good! .. 
its percision in the case of 0 class is 100% .. which means that it correctly classified those cases .. however, it fails to correctly classify two values from the 1 class! and obtain and percision of 98%

You can try to use another kernal type and compare its results with the linear kernal to check if the model is performing well or not!